docs/developer/decision-records/2025-02-06-hashicorp-vault-authentication-refactor/README.md

```diff
@@ -0,0 +1,77 @@
+
+## Rationale
+
+The HashiCorp Vault supports multiple different authentication methods, but the current implementation supports
+only one and provides no way to customize or exchange authentication methods. The full list of supported
+methods can be found in the [HashiCorp Vault Documentation](https://developer.hashicorp.com/vault/docs/auth). Some examples are:
+
+* [Token auth method](https://developer.hashicorp.com/vault/docs/auth/token) (already supported)
+* [Kubernetes auth method](https://developer.hashicorp.com/vault/docs/auth/kubernetes)
+* [AppRole auth method](https://developer.hashicorp.com/vault/docs/auth/approle)
+
+## Approach
+
+The refactoring will be limited to the `vault-hashicorp` module.
+
+The common denominator among all authentication methods is the ability to provide a `client_token` for
+authentication against the vault. Therefore, obtaining this token will be moved to an interface. 
+
+```java
+public interface HashicorpVaultTokenProvider {
+    String vaultToken();
+}
+```
+
+The [token auth method](https://developer.hashicorp.com/vault/docs/auth/token) is already supported by the current implementation.
+The respective code will be moved to a class implementing `HashicorpVaultTokenProvider`, and this
+class will be the default implementation provided through the `HashicorpVaultExtension`.
+
+### Services accessing the vault
+
+Any service that needs to access the vault will be refactored to use the `HashicorpVaultTokenProvider` for obtaining
+the token. Namely, affected services are:
+
+* `HashicorpVault` (secure key-value store)
+* `HashicorpVaultSignatureService` (signing service)
+* `HashicorpVaultHealthService` (health check service)
+
+All of these services currently get the token from the `HashicorpVaultSettings`, which will be replaced by a call to
+the `HashicorpVaultTokenProvider`.
+
+### HashicorpVaultHealthService
+
+In addition, the `HashicorpVaultHealthService` currently stores the headers for all requests to the vault in a
+class variable, which is populated once upon service instantiation. As the token, and thus the headers,
+may not be static for all authentication methods, the variable will be removed. Instead, the token will be fetched from
+the `HashicorpVaultTokenProvider` and the headers will be generated for every request. The token will therefore be
+added as a parameter to the `getHeaders` method.
+
+```java
+@NotNull
+private Headers getHeaders(String token) {
+    var headersBuilder = new Headers.Builder().add(VAULT_REQUEST_HEADER, Boolean.toString(true));
+    headersBuilder.add(VAULT_TOKEN_HEADER, token);
+    return headersBuilder.build();
+}
+```
+
+The `HashicorpVaultHealthService` also provides methods for checking whether a token can be renewed as well as for
+renewing the token. As these features are not available for all authentication methods, this functionality will be
+moved to the `HashicorpVaultTokenRenewTask`.
```

Comment:
for reference: #4751

Reply:
Removed the statement that it will be moved to the `HashicorpVaultTokenRenewTask` in case you already planned to refactor it in a different way, and added a note that this part of the refactoring is already planned for better separation of concerns.